### PR TITLE
Feature: Autoupdate on all platforms, here: WinSparkle

### PR DESCRIPTION
**Pull request by purejava**
This is the first part to provide Autoupdate for Cryptomator.

This PR requires a PR in integrations-win: https://github.com/cryptomator/integrations-win/pull/16

### CODE REVIEW

**Comment by infeo** [comment]
Just for clarifiaction:
> This PR requires a PR in integrations-win: https://github.com/cryptomator/integrations-win/pull/16

No, it is the other way round. This is the "root" PR, since it provides the service interface and attached classes. Providing an implementation is optional, hence the linked PR requires this one.

File: src/main/java/org/cryptomator/integrations/autoupdate/AutoUpdateProvider.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	 * Perform cleanups on the autoupdate features when the application exits.
+	 */
+	void cleanUpAutoUpdate();
+
+	/**
+	 * Enable autoupdate features for the applcation.
+	 *
+	 * @return True, in case enabling the features was successful, false otherwise.
+	 */
+	boolean enableAutoUpdate();
```
Comment (infeo) [inline src/main/java/org/cryptomator/integrations/autoupdate/AutoUpdateProvider.java:38]: Only returning a boolean does not give a deeper insight if enabling it it fails. I suggest to refactor the method:
```java
void enableAutoUpdate() throws AutoUpdateException;
```
Calling it multiple times when already enabled should not throw an exception (aka making the function idempotent).

The documentation should be updated also then.

File: src/main/java/org/cryptomator/integrations/autoupdate/AutoUpdateProvider.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+	/**
+	 * Disable autoupdate features for the applcation.
+	 *
+	 * @return True, in case disabling the features was successful, false otherwise.
+	 */
+	boolean disableAutoUpdate();
```
Comment (infeo) [inline src/main/java/org/cryptomator/integrations/autoupdate/AutoUpdateProvider.java:45]: See comment for `enableAutoUpdate`

File: pom.xml [hunk 3]
```diff
@@ -5,7 +5,7 @@
     <modelVersion>4.0.0</modelVersion>
     <groupId>org.cryptomator</groupId>
     <artifactId>integrations-api</artifactId>
-    <version>1.2.0</version>
+    <version>1.2.0-SNAPSHOT</version>
```
Comment (infeo) [inline pom.xml:8]: Revert this change.

**Review by infeo — CHANGES_REQUESTED** [review]
I don't fully understand the autoupdate logic: Why do two methods exist? (init and cleanup)

When is the autoupdate triggered? Or the check for it? What cleanups are needed for autoupdate when the app exits?

**Comment by purejava** [comment]
> I don't fully understand the autoupdate logic: Why do two methods exist? (init and cleanup)
> 
> When is the autoupdate triggered? Or the check for it? What cleanups are needed for autoupdate when the app exits?

You can see (Win)Sparkle as an add-on: you init it and the handling is done by the add-on library, see https://github.com/vslavik/winsparkle/wiki/Basic-Setup

File: src/main/java/org/cryptomator/integrations/autoupdate/AutoUpdateProvider.java (hunk 1 above)
Comment (purejava) [reply src/main/java/org/cryptomator/integrations/autoupdate/AutoUpdateProvider.java:38]: You are right, I'll change that.

File: pom.xml (hunk 3 above)
Comment (purejava) [reply pom.xml:8]: Ok.

**Comment by infeo** [comment]
> You can see (Win)Sparkle as an add-on: you init it and the handling is done by the add-on library, see https://github.com/vslavik/winsparkle/wiki/Basic-Setup

Ah, got it, thank you for the link. Does Sparkle and AppimageUpdate do work in a similar way or fit into this framework?

When i looked at the WinSparkle api (https://github.com/vslavik/winsparkle/blob/7ca9b991b632fa191d07ed18cd61a2f55a4c1675/include/winsparkle.h#L565), I found a "manual usage" section. I assume WinSparkle and the other implementations work also automatically? (aka calling `init()` checks for updates, shows some UI and on request perform the update)

**Comment by purejava** [comment]
> Ah, got it, thank you for the link. Does Sparkle and AppimageUpdate do work in a similar way or fit into this framework?
> 

WinSparkle is oriented at Sparkle and adapted for Windows. AppImageUpdate is different. It has a limited functionality compared to the other two mentioned before. AppImageUpdates capabilities can be found [here](https://github.com/AppImageCommunity/AppImageUpdate/blob/af298b7aebfab4bcee7490a86f0471cc22db248b/src/cli/main.cpp#L23-L30):
```bash
{"help", {"-h", "--help"}, "Display this help text."},
{"version", {"-V", "--version"}, "Display version information."},
{"describe", {"-d", "--describe"}, "Parse and describe AppImage and its update information and exit."},
{"checkForUpdate", {"-j", "--check-for-update"}, "Check for update. Exits with code 1 if changes are available, 0 if there are not,"
                                                         "other non-zero code in case of errors."},
{"overwriteOldFile", {"-O", "--overwrite"}, "Overwrite existing file. If not specified, a new file will be created, and the old one will remain untouched."},
{"removeOldFile", {"-r", "--remove-old"}, "Remove old AppImage after successful update."},
{"selfUpdate", {"--self-update"}, "Update this AppImage."},
```

So the idea is to bundle AppImageUpdate into our AppImage, so that a newer version of our AppImage can be downloaded next to the running one when an update got available, the running one is ended and the new one is started. I'm on the way to implement this.

> When i looked at the WinSparkle api (https://github.com/vslavik/winsparkle/blob/7ca9b991b632fa191d07ed18cd61a2f55a4c1675/include/winsparkle.h#L565), I found a "manual usage" section. I assume WinSparkle and the other implementations work also automatically? (aka calling `init()` checks for updates, shows some UI and on request perform the update)

This method is intended to trigger checking for an update manually from the UI. This PR implements WinSparkle "the common way", e.g. it pops up on the second start of Cryptomator and in case you enable it, you get a prompt, when a new version for Cryptomator for Windows gets available and you can replace your currently installed Cryptomator installation with the latest one available. WinSparkle does all that for you.
